README.md
[![npm version](https://img.shields.io/npm/v/full-icu.svg)](https://www.npmjs.com/package/full-icu)
[![npm version](https://img.shields.io/npm/v/luxon.svg)](https://www.npmjs.com/package/luxon)
[![GitHub issues](https://img.shields.io/github/issues/sourceforge807/mmm-efa-departures)](https://github.com/sourceforge807/mmm-efa-departures/issues)
[![GitHub forks](https://img.shields.io/github/forks/sourceforge807/mmm-efa-departures)](https://github.com/sourceforge807/mmm-efa-departures/network)
[![GitHub stars](https://img.shields.io/github/stars/sourceforge807/mmm-efa-departures)](https://github.com/sourceforge807/mmm-efa-departures/stargazers)
[![GitHub license](https://img.shields.io/github/license/sourceforge807/mmm-efa-departures?style=plastic)](https://github.com/sourceforge807/mmm-efa-departures/blob/master/LICENSE)
# MMM-EFA-departures
MagicMirror² module to show departures for public transport stations using the EFA system. It will work for all Traffic Companys wich use this service.
This is a fork from https://github.com/Dom1n1c/MMM-EFA-departures.
Thanks for the incredible basics.

- [Installing the module](#installing-the-module)
- [Updating the module](#updating-the-module)
- [Dependencies](#dependencies)
- [Languages](#languages)
- [Prerequisite](#prerequisite)
- [Example Configuration](#example-configuration)
- [Getting Station and Line IDs](#getting-station-and-line-ids)
- [Screenshot](#screenshot)
- [Confguration Options](#confguration-options)
- [Licence](#licence)
- [Versions](#versions)

## Installing the module

To install the module, assuming you have MagicMirror installed with the default configuration:

```shell
cd ~/MagicMirror/modules
git clone https://github.com/sourceforge807/MMM-EFA-departures.git
cd MMM-EFA-departures/
npm install
```

## Updating the module

Navigate to your module folder:
```shell
cd ~/MagicMirror/modules/MMM-EFA-departures
```
use command to pull new git from:
```shell
git pull
```
If you are experiencing the following message “your local changes to the following files would be overwritten by merge” because you have done local edits run this before (NOTE! backup the files you have edited in case you will need to pull some edits back, for exampel the CSS file if you have customized the look of the module:
```shell
git reset --hard
```

## Dependencies
* luxon<br>
Because moment.js is in maintenance mode I change to luxon.
Install luxon in modules/MMM-EFA-departures with:
```shell
npm install --save luxon
```
* node.js<br>
You also need node.js 13+ vor full ICU support. For older versions of node.js you need this guide: https://moment.github.io/luxon/#/install
hint: run
```shell
npm run rebuild
```
after update the node.js
* npm<br>
* [request](https://www.npmjs.com/package/request)<br>

## Languages
As of version 0.1.0, MMM-EFA-departures features language support for `German (de)` and `English (en)` mirrors.

## Prerequisite
A working installation of [MagicMirror<sup>2</sup>](https://github.com/MichMich/MagicMirror)

## Example Configuration
**for Hannover HBF/Main Station:**
```
...
{
	disabled: false,
	module: "MMM-EFA-departures",
	header: "N&auml;chste Bahn",
	position: "top_right",
	config:
	{
		efaUrl: "http://efa107.efa.de/efaws2/default/XML_DM_REQUEST",
		stopID: "25000031",
		stopName: "MMM-EFA is loading", 		//initial module name
		lines: ["all"], 				//lines: ['DDB:92E01: :H','DDB:92E01: :R'], would only show the line S1 in both directions; ['stop'] is a different option; if you will use specific lines use the stateless field in the result from the search script (except the last field :j21)
		reload: 60000, 					//interval in ms (60000=60s)
		realDepTime: true, 				//use real-time data; if there is no realtime data, you have italic formatet messages
		toggleDepTime: true, 				//Toggle relative/absolute time
		toggleDepTimePerReload: 6, 			//Every 6 seconds
		fade: true, 					//fade brightness
		fadePoint: 0.25, 				//Start on 1/4th of the list. (1/maxDepartures would be ideal)
		maxDepartures: 10, 				//maximum amount of departures displayed
		shortenMessage: 12, 				//false or a number
		departureReplace:
		{
			"Hannover" : "H.-",
			"Hildesheim" : "HI.-",
			"Langenhagen" : "Lgh.-"
		}, //wich names will be replaced and shorten?
		lineInfos: true,				//show additional line info

 	}
}
...
```

### Getting Station and Line IDs  
you can get the information by viewing the source (in Chrome for instance) of the EFA-Page  
**view-source:http://efa107.efa.de/efaws2/default/XML_DM_REQUEST**  
To extract and search the line info, use the searchStation.sh script. 

## Screenshot
![grafik](https://user-images.githubusercontent.com/30810603/132315707-edf2e469-0bba-4ed2-8ea4-be5c8114b2f5.png)
![grafik](https://user-images.githubusercontent.com/30810603/132316131-7b1c753e-c612-46bb-88eb-eb5072f9c97f.png)

## Confguration Options

| Name           | Optional | Description |
|----------------|---------------|---------------------------------|
| `efaUrl`|**REQUIRED**| Url to the efa page for the XML_DM_REQUEST. Do not change this url!! <br><br> **Default value:** _http://efa107.efa.de/efaws2/default/XML_DM_REQUEST_ |
| `stopID`|**REQUIRED**| stopID offered by the provider or using the searchStation.sh script. |
| `stopName`|X| initial module name. You can change it...or not.<br><br>**Possible values:** [text] <br> **Default value:** `MMM-EFA is loading`
| `lines`|X| Whitch lines needs to be fetched? If you will use spicific lines, use the stateless field in the result from the search script (except the last field :j21)<br><br> **Possible values:** `['all']`,`['stop']` or maybe `['DDB:92E01: :H','DDB:92E01: :R']`  <br> **Default value:** `['stop']`
| `reload`|X| Reloadintervall in seconds. <br><br> **Possible values:**  <br> **Default value:** `60`
| `realDepTime`|X| Use realtimedata or not. If there is no realtime data available, it will be shown in italic.<br><br> **Possible values:** `true` or `false` <br> **Default value:** `false`
| `toggleDepTime`|X| Toggle between relativ/ absolute time.<br><br> **Possible values:** `true` or `false` <br> **Default value:** `true`
| `toggleDepTimePerReload`|X| Intervall for toggeling the departuretime. <br><br> **Possible values:** `0` - `?` <br> **Default value:** `6` (seconds) |
| `fade`|X| Fade the departures?<br><br> **Possible values:** `true` or `false` <br> **Default value:** `true`
| `fadePoint`|X| Where to start fade? <br><br> **Possible values:** `0` (top of the list) - `1` (bottom of list) <br> **Default value:** `0.25`
| `maxDepartures`|X| How many departures will be shown? <br><br>**Possible values:** `0` - `?` <br> **Default value:** `4` |
| `shortenMessage`|X| Shortens the name of the targets. <br><br> **Possible values:** `0` - `?` <br> **Default value:** `12` |
| `language`|X| Select the Language. <br><br> **Possible values:** `de`, `en` <br> **Default value:** `de` |
| `departureReplace`|X| Names wich will be replaced with something you want an the departures will be shorten. <br><br> **Possible values:** `[text]` <br> **Default value:** `{"Hannover" : "H.-", "Hildesheim" : "HI.-", "Langenhagen" : "Lgh.-"}` |
| `linieInfos`|X| Show additional line info. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `true` |
| `stopInfos`|X| Show additional stop info. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `false` |
| `showDelay`|X| Show the delay, Delays of up to 5 minutes are shown in bold. Delays between 5 and 10 minutes are shown in bold and orange. Delays of 10 minutes or more are displayed in bold and red.  <br><br> **Possible values:** `true` or `false` <br> **Default value:** `true` |
| `showTrainColor`|X| Shows the train type in color. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `true` |
| `colorStadtbahn`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#8E44AD` |
| `colorSBahn`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#CB4335` |
| `colorRBahn`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#17A589` |
| `colorIC`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#6495ED` |
| `colorICE`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#F1C40F` |
| `colorEC`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#A04C1A` |
| `colorBus`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#0404B4` |
| `colorErsatzverkehr`|X| Color of the type. <br><br> **Possible values:** All of CSS.supports() colors. <br> **Default value:** `#4B4040` |
| `showServingLineDelay`|X| Shows the servingLine delay in minutes. <br><br> **Possible values:** `true` or `false` <br> **Default value:** `false` |

## Licence
MIT License

Copyright (c) 2021 sourceforge807 (https://github.com/sourceforge807/)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

## Versions
| Version           | Description |
|----------------|---------------|
| `0.0.1`|initial|
| `0.1.0`|Some bugfixes an new features.|


changelog.md
# MMM-EFA-departures changelog
This module adheres to [Semantic Versioning](http://semver.org/).

## [0.1.0] - 2021-09-07

### Added
- `maxLinesOverall` to limit the max rows. It overrules the maxDepartures
- `departureTransform()` to replace not required text in departures saved in `departureReplace: {}`
- `shortenMessage()` to short the departures by using `shortenMessage`. used by `departureTransform()`
- `lineInfos` option to show additional line infos
- `stopInfos` option to show additional stop infos
- `showTrainColor` option for show collored train types an define there color by `colorStadtbahn`,`colorSBahn`,`colorRBahn`,`colorIC`,`colorICE`,`colorEC`,`colorBus`,`colorErsatzverkehr`,
- `showDelay` option to show the delay as a `style=""` element (bold an bold + color)
- Humanize seconds to "now"
- `sanitizeNumbers()` in the config
- `checkColor()` for the colored train type
- Check if is the `stopID` and `efaURL` is empty
- Add `text-decoration` to `line-through` if the trip ist cancelled
- Add some css attributes
- `servingLineDelay` option to show the delay of the servinLine in minutes
- New realtimeStatus `EXTRA_TRIP`

### Fixed
- Fixed some smaller bugs
- Fixed the translations (de, en)
- Fixed the css toggle attributes
- Sort departure by countdown

### Changed
- Remove moment.js and use luxon instead

## [0.0.1] - 2021-08-24 
### Initial release


node_helper.js
const request = require('request');
const moment = require('moment');
const NodeHelper = require("node_helper");

module.exports = NodeHelper.create({

    start: function() {
        console.log("Starting node helper for: " + this.name);
		this.config = null;
    },

    socketNotificationReceived: function(notification, payload) {
		var self = this;
        if(notification === 'CONFIG'){
            this.config = payload;

			var efa_url = this.config.efaUrl;
				efa_url += '?sessionID=0';				
				efa_url += '&requestID=0';
				efa_url += '&language=de';
				efa_url += '&useRealtime=' + this.config.realDepTime;
				efa_url += '&mode=direct';
				efa_url += '&dmLineSelectionAll=1';
				efa_url += '&name_dm=' + this.config.stopID;
				efa_url += '&type_dm=stop';
				efa_url += '&line=' + this.config.lines.join('&line=');
				efa_url += '&outputFormat=json';
				efa_url += '&limit=' + this.config.maxDepartures;
				efa_url += '&itdTime=' + moment().format('HHmm');
				efa_url += '&itdDate=' + moment().format('YYYYMMDD');
				efa_url += '&outputEncoding=UTF-8';
				efa_url += '&inputEncoding=UTF-8';
				efa_url += '&mId=efa_www';



			console.log("Efa url '" + efa_url + "' for stop '" + this.config.stopID + "' created.");

            this.getData(efa_url, this.config.stopID);
        }
    },

    getData: function(options, stopID) {

		var error = 0;

        request(options, (error, response, body) => {

    		if(!response || response === undefined || response.statusCode != 200) {
      			console.log("Error getting response");
				error = 1;
    		}

            if (!error || response.statusCode === 200) {

				console.log("Socket notification TRAMS with '" + this.config.stopID + "' status code ==> '" + response.statusCode + "'");

                this.sendSocketNotification("TRAMS" + stopID, JSON.parse(body));
            } else if (response === undefined) {
                console.log("Error getting tram connections; Response undefined");
            } else {
				console.log("Error getting tram connections " + response.statusCode); // ${response.statusCode} ??
			}
        });
    }
});


MMM-EFA-departures.js
/* Magic Mirror
 * Module: MMM-EFA-departures
 *
 * By gefangenimnetz / https://github.com/gefangenimnetz/MMM-EFA-departures
 * MIT Licensed.
 * Forked and modified by sourceforge807 / https://github.com/sourceforge807
 * 
 * v0.1.0
 */

Module.register("MMM-EFA-departures", {
	requiresVersion: '2.1.0',
    defaults: {
		efaUrl: "http://efa107.efa.de/efaws2/default/XML_DM_REQUEST",
		stopID: "25000031",							//stopID offered by the provider (Hannover HBF in this case)
		stopName: "MMM-EFA is loading",						//initial module name
		lines: ['stop'],							//lines: ['DDB:92E01: :H','DDB:92E01: :R'], would only show the line S1 in both directions; ['all'] is a different option; if you will use specific lines use the stateless field in the result from the search script (except the last field :j21)
		reload: 60000,								//interval in ms (60000=60s)
		realDepTime: false,							//use real-time data
		toggleDepTime: true,							//Toggle relative/absolute time
		toggleDepTimePerReload: 6,						//Every 6 seconds
		fade: true,								//fade brightness
		fadePoint: 0.25,							//Start on 1/4th of the list. (1/maxDepartures would be ideal)
		maxDepartures: 4,							//maximum amount of departures displayed
		maxLinesOverall : 10,							//include text messages; this overrides the max departures lines
		shortenMessage: 12, 							//false or a number
		language: "ru",								//select de or en
		departureReplace: {"Hannover" : "H.-", "Hildesheim" : "HI.-"},
		lineInfos: true,							//show additional line info
		stopInfos: false,							//show additional stop info
		showTrainColor: true,
		colorStadtbahn: "#8E44AD",
		colorSBahn: "#CB4335",
		colorRBahn: "#17A589",
		colorIC: "#6495ED",
		colorICE: "#F1C40F",
		colorEC: "#A04C1A",
		colorBus: "#0404B4",
		colorErsatzverkehr: "#4B4040",
		showDelay: true,
	    	showServingLineDelay: false,
    },

    start: function () {
		var self = this;
		Log.info("Starting module: " + this.name);

		this.sendSocketNotification("CONFIG", self.config);
		setInterval(function()
		{
			self.sendSocketNotification("CONFIG", self.config);
		}, this.config.reload);

		this.sanitizeNumbers([
			"stopID",
			"reload",
			"toggleDepTimePerReload",
			"maxDepartures",
			"shortenMessage"
		]);

		this.checkColor([
			"colorStadtbahn",
			"colorSBahn",
			"colorRBahn",
			"colorIC",
			"colorICE",
			"colorEC",
			"colorBus",
			"colorErsatzverkehr"
		]);
	},

	getStyles: function ()
	{
		return ["MMM-EFA-departures.css"];
	},

	getScripts: function()
	{
		return ["classie.js", this.file("/node_modules/luxon/build/global/luxon.min.js")];
	},

	getTranslations: function ()
	{
		return {
			de: "translations/de.json",
			en: "translations/en.json"
		};
	},

	socketNotificationReceived: function (notification, payload)
	{
		if (notification === "TRAMS" + this.config.stopID)
		{
			this.efa_data = payload;
			this.config.stopName = this.translate("FROM") + payload.dm.input.input;
			this.updateDom(500);           
		}
	},
                    
	getDom: function ()
	{
		var wrapper = document.createElement("div");

		if ( this.config.stopID === "" )
		{
			wrapper.innerHTML = this.translate("NOSTOPID") + this.name + this.translate("DOT");
			wrapper.className = "dimmed light small";
			return wrapper;
		}

		if ( this.config.efaUrl === "" )
		{
			wrapper.innerHTML = this.translate("NOEFAURL") + this.name + this.translate("DOT");
			wrapper.className = "dimmed light small";
			return wrapper;
		}

		var header = document.createElement("header");
		header.innerHTML = this.config.stopName;
		wrapper.appendChild(header);

		this.loaded = true;

		if ( this.loaded === false )
		{
			var text = document.createElement("div");
			text.innerHTML = this.translate("LOADING");
			text.className = "dimmed light small";
			wrapper.appendChild(text);
		}
		else if ( !this.efa_data )
		{
			var text = document.createElement("div");
			text.innerHTML = this.translate("NODATA");
			text.className = "dimmed light small";
			wrapper.appendChild(text);
		}
		else
		{
			var departuresUL = document.createElement("ul");
			departuresUL.className = 'small';
			var departures = this.efa_data.departureList;
			var counter = 1;



			// classie ersetzen mit https://www.w3schools.com/howto/howto_js_toggle_class.asp ?!?!
			if ( this.config.toggleDepTime )
			{
				window.clearInterval(this.toggleTimeInt);
				this.toggleTimeInt = window.setInterval(function()
				{
					classie.toggle(departuresUL, 'departures__departure--show-time');
				}, (this.config.reload / this.config.toggleDepTimePerReload));
			}

			//sort by countdown
			function compare(a, b)
			{
				if (parseInt(a.countdown) > parseInt(b.countdown)) return 1;
				if (parseInt(b.countdown) > parseInt(a.countdown)) return -1;
				return 0;
			}
			departures.sort(compare);

			for ( var d in departures )
			{
				if ( counter > this.config.maxLinesOverall )
				{
					break;
				}

				var departuresLI = document.createElement("li");
				departuresLI.className = 'departures__departure';

				if ( this.config.realDepTime === true && departures[d].servingLine.realtime === '1' && departures[d].hasOwnProperty('realDateTime') === true )
				{
					var departureTime = luxon.DateTime.local(
														parseInt(departures[d].realDateTime.year),
														parseInt(departures[d].realDateTime.month),
														parseInt(departures[d].realDateTime.day),
														parseInt(departures[d].realDateTime.hour),
														parseInt(departures[d].realDateTime.minute)
													);
				}
				else
				{
					var departureTime = luxon.DateTime.local(
														parseInt(departures[d].dateTime.year),
														parseInt(departures[d].dateTime.month),
														parseInt(departures[d].dateTime.day),
														parseInt(departures[d].dateTime.hour),
														parseInt(departures[d].dateTime.minute)
													);
				}

				var departureTimeRelative = departureTime.setLocale(this.config.language).toRelative({ base: luxon.DateTime.now(), style: "short" });
				var departureTimeAbsolute = departureTime.toLocaleString(luxon.DateTime.TIME_24_SIMPLE);

				// humanize seconds
				var diff = Math.round((departureTime- luxon.DateTime.now()) / (1000*60));
				if ( (diff <= 1 && diff >= 0) || (diff >= -1 && diff <= 0) )
				{
					departureTimeRelative = this.translate("NOW");
				}

				// slicing and transform departures
				var departure = departures[d].servingLine.direction;
				if ( this.config.shortenMessage && departure.length > this.config.shortenMessage )
				{
					departure = this.departureTransform(departure);
				}

				// slicing for long ice number + names
				var servingLineNumber = departures[d].servingLine.number;
				if ( servingLineNumber.length >= 9 )
				{
					servingLineNumber = servingLineNumber.slice(0, 8);
				}

				var backgroundColor = "";
				var tripText = "";
				var tripCancelled = "";
				var styleTrainNumber = '';
				var styleTrainName = '';

				if ( departures[d].hasOwnProperty('realtimeStatus') === true ){
					switch (departures[d].realtimeStatus) {
						case 'TRIP_CANCELLED':
							backgroundColor = "backgroundPurple";
							tripText = this.translate("CANCELLED");
							tripCancelled = 'text-decoration:line-through\;'; //collision with style.opacity -> fading ??
							break;
						case 'EXTRA_TRIP':
							backgroundColor = "backgroundPurple";
							tripText = this.translate("EXTRA_TRIP");
							tripCancelled = 'text-decoration:underline red\;';
							break;
						default:
							backgroundColor = "backgroundWhite";
							tripText= this.translate("UNKNOWN");
							tripCancelled = 'text-decoration:blink\;';
					}
				}

				var trainColor = '';
				if ( this.config.showTrainColor === true )
				{
					var trainColor = this.getTrainColor(departures[d].servingLine.name, departures[d].servingLine.trainType);
				}
				else
				{
					var trainColor = "";
				}

				// shows the delay as a style
				var delay = '';
				var servingLineDelay = '';
				if ( this.config.showDelay === true && departures[d].servingLine.hasOwnProperty('delay') === true )
				{
					if ( departures[d].servingLine.delay > 0 && departures[d].servingLine.delay <= 5 )
					{
						delay = 'font-weight: bold\;';
					}
					else if ( departures[d].servingLine.delay > 5 && departures[d].servingLine.delay <= 10 )
					{
						delay = 'font-weight: bold; color: orange\;';
					}
					else if ( departures[d].servingLine.delay > 10 )
					{
						delay = 'font-weight: bold; color: red\;';
					}
					
					var sign = '';
					if ( departures[d].servingLine.delay > 0 && this.config.showServingLineDelay === true )
					{
						sign = '+';
						servingLineDelay = ' (' + sign + departures[d].servingLine.delay + ') ';
					}
					else if ( departures[d].servingLine.delay < 0 && this.config.showServingLineDelay === true )
					{
						sign = '-';
						servingLineDelay = ' (' + sign + departures[d].servingLine.delay + ')';
					}
					else
					{
						sign = '';
						servingLineDelay = '';
					}


				}

				styleTrainNumber = 'style="' + tripCancelled + trainColor + '"';
				styleTrainName = 'style="' + tripCancelled + delay + '"';

				if ( this.config.realDepTime === true && departures[d].servingLine.realtime === '1' )
				{
					departuresLI.innerHTML = '<span class="departures__departure__line__realtime xsmall" ' + styleTrainNumber + '>' + servingLineNumber + '</span><span class="departures__departure__direction__realtime small' + backgroundColor + '" ' + styleTrainName + '>' + departure + tripText + '&nbsp;&nbsp;</span><span class="departures__departure__time__realtime-relative small bright" ' + styleTrainName + '>' + departureTimeRelative + '</span><span class="departures__departure__time__realtime-clock small bright" ' + styleTrainName + '>' + departureTimeAbsolute + '</span>';
					counter++;
				}
				else
				{
					departuresLI.innerHTML = '<span class="departures__departure__line xsmall" ' + styleTrainNumber + '>' + servingLineNumber + '</span><span class="departures__departure__direction small ' + backgroundColor + '" ' + styleTrainName + '>' + departure + tripText + '&nbsp;&nbsp;</span><span class="departures__departure__time-relative small bright" ' + styleTrainName + '>' + departureTimeRelative + '</span><span class="departures__departure__time-clock small bright" ' + styleTrainName + '>' + departureTimeAbsolute + '</span>';
					counter++;
				}


				if ( this.config.fade && this.config.fadePoint < 1 )
				{
					if ( this.config.fadePoint < 0 )
					{
						this.config.fadePoint = 0;
					}
					var startingPoint = departures.length * this.config.fadePoint;
					var steps = departures.length - startingPoint;
					if ( d >= startingPoint )
					{
						var currentStep = d - startingPoint;
						departuresLI.style.opacity = 1 - (1 / steps * currentStep);
					}
				}
				departuresUL.appendChild(departuresLI);

				// bahnen
				if ( departures[d].hasOwnProperty('lineInfos') === true && this.config.lineInfos === true && departures[d].lineInfos )
				{
					var lineInfoLI = document.createElement("li");
					lineInfoLI.className = 'marquee';
					lineInfoLI.innerHTML += '<span class="small" id="' + d + '">' + departures[d].lineInfos.lineInfo.infoText.subtitle + '</span>';
					lineInfoLI.style.opacity = departuresLI.style.opacity;
					departuresUL.appendChild(lineInfoLI);
					counter++;
				}


				// busse
				if ( departures[d].hasOwnProperty('stopInfos') === true && this.config.stopInfos === true && departures[d].stopInfos )
				{
					for ( var e in departures[d].stopInfos )
					{
						var stopInfoLI = document.createElement("li");
						stopInfoLI.className = 'marquee';
						stopInfoLI.innerHTML += '<span class="small" id="' + d + '">' + departures[d].stopInfos[e].infoText.subtitle + '</span>';
						stopInfoLI.style.opacity = departuresLI.style.opacity;
						departuresUL.appendChild(stopInfoLI);
						counter++;
					}
					
				}
			}
			wrapper.appendChild(departuresUL);
		}
		return wrapper;
	},

	/* shorten(string, maxLength)
	 * Shortens a string if it's longer than maxLength.
	 * Adds an ellipsis to the end.
	 *
	 * argument string string - The string to shorten.
	 * argument maxLength number - The max length of the string.
	 *
	 * return string - The shortened string.
	 */
	shorten: function (string, maxLength)
	{
		if ( string.length > maxLength )
		{
			return string.slice(0, maxLength) + "&hellip;"; // &#8230;
		}

		return string;
	},


	/* departureTransform(departure)
	 * Transforms the departure for usage.
	 * Replaces parts of the text as defined in config.departureReplace.
	 * Shortens departure based on config.shortenMessage
	 *
	 * argument departure string - The departure to transform.
	 *
	 * return string - The transformed departure.
	 */
	departureTransform: function (departure)
	{
		for ( var needle in this.config.departureReplace )
		{
			var replacement = this.config.departureReplace[needle];

			var regParts = needle.match(/^\/(.+)\/([gim]*)$/);
			if ( regParts )
			{
			  // the parsed pattern is a regexp.
			  needle = new RegExp(regParts[1], regParts[2]);
			}

			departure = departure.replace(needle, replacement);
		}
		departure = this.shorten(departure, this.config.shortenMessage);
		return departure;
	},

	/*
	 * For any config parameters that are expected as integers, this
	 * routine ensures they are numbers, and if they cannot be
	 * converted to integers, then the module defaults are used.
	 *
	 * argument key int - The keys to sanitize.
	 *
	 * return key - The sanitized or the default key.
	 */
	sanitizeNumbers: function(keys)
	{

		var self = this;
		keys.forEach(function(key)
		{
			if ( isNaN(parseInt(self.config[key])) )
			{
				self.config[key] = self.defaults[key];
			}
			else
			{
				self.config[key] = parseInt(self.config[key]);
			}
		});
	},

	checkColor: function(keys)
	{
		var self = this;
		keys.forEach(function(key)
		{
			if ( !CSS.supports("color", self.config[key]) )
			{
				self.config[key] = self.defaults[key];
			}
		});
	},

	getTrainColor: function(name, type)
	{
		var style = '';

		if ( type == "ICE" || type == "IC" || type == "NJ" )
		{
			name = type;
		}

		switch (name) {
			case 'Stadtbahn':
				style = 'border-color:' + this.config.colorStadtbahn + '; border-style: solid; border-radius: 6px\;';
				break;
			case 'S-Bahn':
				style = 'border-color:' + this.config.colorSBahn + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			case 'R-Bahn':
				style = 'border-color:' + this.config.colorRBahn + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			case 'NJ':
			case 'nightjet':
			case 'InterCity':
				style = 'border-color:' + this.config.colorIC + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			case 'ICE':
			case 'InterCityExpress':
				style = 'border-color:' + this.config.colorICE + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			case 'EC':
			case 'EuroCity':
				style = 'border-color:' + this.config.colorEC + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			case 'Bus':
				style = 'border-color:' + this.config.colorBus + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			case 'Ersatzverkehr':
				style = 'border-color:' + this.config.colorErsatzverkehr + '\; border-style: solid\; border-radius: 6px\;"';
				break;
			default:
				style = 'border-color:black; border-style: solid\; border-radius: 6px\;"';
		}

		return style;
	}
});
